### ISSUE TITLE
Flake fragments should prioritise the top-most attribute.

### ISSUE THREAD

**Issue by L-as** [NixOS/nix#5651]
**Describe the bug**

Doing `nix build x#y`, `nix eval x#y` or similar, will cause Nix to evaluate `packages.<system>.y` *before* `y`.
This is problematic in my case, because `packages` uses IFD, whereas `y` does not.
I want to be able to evaluate `y`, which does not use IFD, with `--no-allow-import-from-derivation`.

**Steps To Reproduce**

1) Have `packages` use IFD.
2) Define top-level attribute `y`.
3) Do `nix --no-allow-import-from-derivation eval .#y`. This should fail. 

**Expected behavior**

Nix should evaluate the top-most attribute before trying the one in `packages` and similar.

**`nix-env --version` output**
`nix (Nix) 2.5pre20211007_844dd90`

**Comment by lilyball** [comment]
Copying https://github.com/NixOS/nix/pull/5657#issuecomment-980458038 here as it really applies to the issue, not the implementation:

---

This is going to make `nixpkgs#lib` refer to the top-level `lib` in the flake instead of the `lib` in the package set. At the very least this means any overlays that update `lib` wouldn't apply to `nixpkgs#lib` after this change; I don't know if there are any other practical differences in this case besides the fact that the top-level `lib` has a `nixosSystem` function. But it's an example of a potentially surprising outcome of this change.

More generally, if `nixpkgs#hello` resolves to `nixpkgs#legacyPackages.aarch64-darwin.hello` then `nixpkgs#foo` should resolve to `nixpkgs#legacyPackages.aarch64-darwin.foo`. I shouldn't have to know the complete set of outputs of a flake to predict where arbitrary fragments resolve to. Also as a reproducibility issue, a flake adding a new top-level output shouldn't break existing flake references (e.g. nixpkgs adding a top-level `hello` output shouldn't break `nixpkgs#hello`).

I feel like perhaps a better approach is to have a separate syntax for "don't apply any prefixes to this fragment". For example, something like `nixpkgs##lib`.

**Comment by regnat** [comment]
> More generally, if nixpkgs#hello resolves to nixpkgs#legacyPackages.aarch64-darwin.hello then nixpkgs#foo should resolve to nixpkgs#legacyPackages.aarch64-darwin.foo. I shouldn't have to know the complete set of outputs of a flake to predict where arbitrary fragments resolve to. Also as a reproducibility issue, a flake adding a new top-level output shouldn't break existing flake references (e.g. nixpkgs adding a top-level hello output shouldn't break nixpkgs#hello).

That’s true, but it also play the other way: if `nixpkgs#foo` resolves to `nixpkgs#foo`, then `nixpkgs#hello` should resolve to `nixpkgs#hello`. Which isn’t the case.

An “interesting” side-effect of the current behavior is that if my flake has both a toplevel `foo` attribute and a `packages.{mySystem}.foo` attribute, then there’s no way (to my knowledge) to `nix eval` the toplevel foo attribute

**Comment by lilyball** [comment]
@regnat Despite my attempt to move the discussion here, it has continued in the PR (see https://github.com/NixOS/nix/pull/5657#issuecomment-980480993).

**Comment by tomberek** [comment]
> I feel like perhaps a better approach is to have a separate syntax for "don't apply any prefixes to this fragment". For example, something like `nixpkgs##lib`.

I'm looking at an approach like this: https://github.com/NixOS/nix/compare/master...flox:nix:disable_attrpath_resolution

If this is preferable I can add tests/docs/etc. It's a bit odd because we're using an invalid URL, but it is also for the unique situations when one wants to express "don't perform automatic attrPath resolution".

**Comment by roberth** [comment]
Why would `eval` implement any rules if it doesn't come with its own flake attribute? `eval --package`, sure, but this is weird.

> URL

We need to stop pretending that they're universal.

**Comment by ysndr** [comment]
Something perhaps more compatible might be "flake#.hello".
It's not a valid nix attrpath but it allows for the attrpath to be parsed as `["", "hello"]` using the existing separator.
An attrpath starting with `""` could then be considered a "rooted" attrpath enforcing fully qualified access while interpreted access is still possible with the paths we know

**Comment by ysndr** [comment]
In fact that already works by adding 
```
"" = self;
```
to the flake output

**Comment by tomberek** [comment]
A lighter approach using `#.`: https://github.com/flox/nix/commit/74639338e6db1d027068aae1d4e5679440fec1e4

Supports `nix eval .#. --apply builtins.attrNames`

Old: https://github.com/flox/nix/commit/e5d3fcb3a0ff9ca3189cbe5ad421b9b9f563c7e6
